Client/Frontend/Browser/BrowserViewController.swift

```diff
@@ -2992,6 +2992,54 @@ extension BrowserViewController: WKUIDelegate {
+                
+                guard let contextHelper = currentTab.getContentScript(name: ContextMenuHelper.name()) as? ContextMenuHelper,
+                let elements = contextHelper.elements else { return nil }
+                
+                if let url = elements.image {
+                    actions.append(UIAction(title: Strings.saveImageActionTitle, identifier: UIAction.Identifier("linkContextMenu.saveImage")) { _ in
+                        self.getData(url) { [weak self] data in
+                            guard let self = self,
+                                  let image = data.isGIF ? UIImage.imageFromGIFDataThreadSafe(data) : UIImage.imageFromDataThreadSafe(data) else { return }
+
+                            UIImageWriteToSavedPhotosAlbum(image, self, #selector(self.image(image:didFinishSavingwithError:contextInfo:)), nil)
+                        }
+                    })
+
+                    actions.append(UIAction(title: Strings.copyImageActionTitle, identifier: UIAction.Identifier("linkContextMenu.copyImage")) { _ in
+                        // put the actual image on the clipboard
+                        // do this asynchronously just in case we're in a low bandwidth situation
+                        let pasteboard = UIPasteboard.general
+                        pasteboard.url = url as URL
+                        let changeCount = pasteboard.changeCount
+                        let application = UIApplication.shared
+                        var taskId: UIBackgroundTaskIdentifier = UIBackgroundTaskIdentifier(rawValue: 0)
+                        taskId = application.beginBackgroundTask(expirationHandler: {
+                            application.endBackgroundTask(taskId)
+                        })
+
+                        URLSession(configuration: .default, delegate: nil, delegateQueue: .main).dataTask(with: url, completionHandler: { data, response, error in
+                            if let response = response as? HTTPURLResponse {
+                                if !(200..<300).contains(response.statusCode) {
+                                    return application.endBackgroundTask(taskId)
+                                }
+                            }
+                                                    
+                            // Only set the image onto the pasteboard if the pasteboard hasn't changed since
+                            // fetching the image; otherwise, in low-bandwidth situations,
+                            // we might be overwriting something that the user has subsequently added.
+                            if changeCount == pasteboard.changeCount, let imageData = data, error == nil {
+                                pasteboard.addImageWithData(imageData, forURL: url)
+                            }
+                            
+                            application.endBackgroundTask(taskId)
+                        }).resume()
+                    })
```

Comment:
Shouldn't we already have the image downloaded on the page to copy?

Reply:
You are right. I will check in detail why downloading again fixes this problem